crates/apollo-compiler/src/name.rs

```diff
@@ -0,0 +1,511 @@
+
+    // The `unsafe impl`s below are sound since `(tag, ptr, len)` represents `UnpackedRepr`
+    const fn assert_send_and_sync<T: Send + Sync>() {}
+    assert_send_and_sync::<(UnpackedRepr, u32, TaggedFileId)>();
+};
+
+unsafe impl Send for Name {}
+
+unsafe impl Sync for Name {}
+
+impl Name {
+    /// Create a new `Name` parsed from the given source location
+    pub fn new_parsed(value: &str, location: NodeLocation) -> Result<Self, InvalidNameError> {
+        Self::check_valid_syntax(value, Some(location))?;
+        let (ptr, len, tag) = Self::parts_from_arc(value);
+        let start_offset = Self::with_location(value, &location);
+        Ok(Self {
+            ptr,
+            len,
+            start_offset,
+            tagged_file_id: TaggedFileId::pack(tag, location.file_id),
+            phantom: PhantomData,
+        })
+    }
+
+    /// Create a new `Name` programatically, not parsed from a source file
+    pub fn new(value: &str) -> Result<Self, InvalidNameError> {
+        Self::check_valid_syntax(value, None)?;
+        let (ptr, len, tag) = Self::parts_from_arc(value);
+        Ok(Self {
+            ptr,
+            len,
+            start_offset: 0,
+            tagged_file_id: TaggedFileId::pack(tag, FileId::NONE),
+            phantom: PhantomData,
+        })
+    }
+
+    /// Create a new static `Name` programatically, not parsed from a source file
+    pub fn new_static(value: &'static str) -> Result<Self, InvalidNameError> {
+        Self::check_valid_syntax(value, None)?;
+        Ok(Self::new_static_unchecked(value))
+    }
+
+    /// Create a new static `Name` programatically, not parsed from a source file,
+    /// without validity checking.
+    ///
+    /// Constructing an invalid name may cause invalid document serialization
+    /// but not memory-safety issues.
+    pub const fn new_static_unchecked(value: &'static str) -> Self {
+        let (ptr, len, tag) = Self::parts_from_static(value);
+        Self {
+            ptr,
+            len,
+            start_offset: 0,
+            tagged_file_id: TaggedFileId::pack(tag, FileId::NONE),
+            phantom: PhantomData,
+        }
+    }
+
+    fn parts_from_arc(value: &str) -> (NonNull<u8>, u32, bool) {
+        let len = Self::new_len(value);
+        let arc = Arc::<str>::from(value);
+        let ptr = Arc::into_raw(arc).cast_mut().cast();
+        // SAFETY: Arc always is non-null
+        let ptr = unsafe { NonNull::new_unchecked(ptr) };
+        (ptr, len, TAG_ARC)
+    }
+
+    const fn parts_from_static(static_str: &'static str) -> (NonNull<u8>, u32, bool) {
+        let len = Self::new_len(static_str);
+        let ptr = static_str.as_ptr().cast_mut();
+        // SAFETY: `&'static str` is always non-null
+        let ptr = unsafe { NonNull::new_unchecked(ptr) };
+        (ptr, len, TAG_STATIC)
+    }
+
+    /// Returns start_offset
+    fn with_location(value: &str, location: &NodeLocation) -> u32 {
```

Comment:
`with_` is usually a constructor method. should we call this `start_offset`?
```suggestion
    fn start_offset(value: &str, location: &NodeLocation) -> u32 {
```

Reply:
This is inlined in a later commit